[Entry] Algorithm: Adding kadane.md into algorithm folder

dakshdeepHERE opened the pull request:
<!--
👋 Hi, thanks for sending a PR to Codecademy Docs! 💖
Please fill out all fields below to ensure your PR is reviewed quickly.

**IMPORTANT**

If you would like to receive credit for your contribution to an entry, make sure to have your Codecademy user profile linked to your GitHub account:

1. Go to your Codecademy dashboard.
2. Click on your profile image in the top-right and then choose "Profile".
3. Then click "Edit Profile".
4. In the "GitHub Username" field, add your username (without the @).
5. Click "Save Changes"

Of course, you can opt not to do this and be listed as an "Anonymous contributor", instead. :)
-->

### Description

<!-- Please write a summary of the change, such as which topic(s) and file(s) that you have edited or created. Please also include relevant motivation and context: -->

### Type of Change

<!--- Please delete or cross off the bullet point(s) that are irrelevant to this PR: -->

- Adding a new entry
- Updating the documentation

### Checklist

<!-- Please check ALL the boxes: -->

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

<!--
Having trouble with the PR checker? Here are some common issues and resolutions:

- verify_formatting is failing
  - run `yarn format path/to/markdown/file.md` or `yarn format:all` and commit the results
- verify_lint is failing
  - same as above
  - if verify_lint is still failing, running `yarn lint` locally should let you know what needs to be changed by hand
- test is failing
  - ensure any new markdown files have a `Title` and `Description` defined in their metadata
  - ensure any new markdown files only contain alphanumerics and dashes in their file names and have the same name as their parent directory
  - if that looks ok, running `yarn test` locally should let you know what the issue is
-->
Fixes #2607

dakshdeepHERE commented:
@HishamT is there any changes required ?

HishamT commented on content/general/concepts/algorithm/terms/euclidean-algorithm/kadane-algorithm.md:
```diff
@@ -0,0 +1,81 @@
+  - 'paths/computer-science'
+---
+
+Kadane's algorithm is an efficient way to find the maximum sum of a contiguous subarray within a given array of numbers. It iterates through the array and keeps track of the current maximum sum and the maximum sum found so far. By comparing the current element with the running sum, it decides whether to continue the subarray or start a new one. The algorithm guarantees a time complexity of O(n),  ...
```
This actually isn't a very clear explanation of Kadane's algorithm. Can you reword it?

HishamT commented on content/general/concepts/algorithm/terms/euclidean-algorithm/kadane-algorithm.md:
```diff
@@ -0,0 +1,81 @@
+
+## Explanation
+
+In problems related to `Kadane's algorithm` an array of numbers will be given, and the goal is to find the biggest sum that can be made using a bunch of consecutive numbers. Skipping any numbers or using them out of order is not allowed.
```
This statement repeats what was said in the previous section. Can you reword it or remove it entirely?

HishamT commented on content/general/concepts/algorithm/terms/euclidean-algorithm/kadane-algorithm.md:
```diff
@@ -0,0 +1,81 @@
+
+**Here's how it can be done:**
+
+- Start with the first number and add it to a running total. Then, move to the next number and decide whether it's better to continue with the current sum or start a new sum from this number. Always aim to keep the largest sum seen so far.
```
The key point that may not be so intuitive is what happens when the we are better off by starting a new sum from the current position. Can you include this in the example? I.e., show a case where it is actually better off by starting a new sum from the current position.

HishamT reviewed (changes requested):
Hey @dakshdeepHERE, thank you so much for your entry on Kadane's algorithm. I did a first review and left some comments. Once they are addressed I will review your entry once again.

dakshdeepHERE commented:
@HishamT I've tried to solve the first two comments. Can you check now. Also for third comment I've a doubt that you can clear for me please? 

**`IDEA`**
Is it permissible to use a LeetCode question as a reference and solve it in the codebyte for this entry using Kadane's algorithm? I can solve the question using the brute force method first and then demonstrate the application of Kadane's algorithm on the same question to showcase the improvement in time complexity.

This is just a idea/suggestion that I got. If we can do this then I can demonstrate the same thing in future algorithm entries too.

HishamT commented:
Hey @dakshdeepHERE, unfortunately Codecademy does not allow content from other websites such as LeetCode, Hackerrank, Codeforces, etc... for internal reasons and conflicts of interest. 

As for the third comment, I wanted you to show the case when it is better for you to start over as opposed to the current sum . e.g., the sum starting at 3 in this array: [1, -2, 3, 4, -5]

dakshdeepHERE commented:
> Hey @dakshdeepHERE, unfortunately Codecademy does not allow content from other websites such as LeetCode, Hackerrank, Codeforces, etc... for internal reasons and conflicts of interest.

Yes this is what I though of don't worry I'm thinking of something good to attach in the codebyte section. So let's say I make a question and then solve it in codebyte? This can work? I'll try to make a question based on kadane's algo. 



> As for the third comment, I wanted you to show the case when it is better for you to start over as opposed to the current sum . e.g., the sum starting at 3 in this array: [1, -2, 3, 4, -5]

Aha I got it now sure this also showcase why we use max(currentSum,0) in Kadane's algo. I will get on it right away. I waited till now to confirm this only. 👍🏻

dakshdeepHERE commented:
Hey @HishamT one more question. What are your views on adding the nested loop method to solve the same problem which is solved by kadane's algo? This will also showcase the huge diff this algo brings in time complexity.

HishamT commented:
Hey! What do you mean by nested loop method?

HishamT commented:
Hey @dakshdeepHERE, do you have any updates regarding this entry?

dakshdeepHERE commented:
> Hey @dakshdeepHERE, what do you mean by nested loop method? 😁

Thats the brute force approach where we iterate through each  combinations and store the sum. In the end we return the largest sum as the answer.

SSwiniarski commented:
@dakshdeepHERE Are you still working on this?

SSwiniarski commented:
@dakshdeepHERE have you given up on this entry? @HishamT how does this look in its current state? Is it ready for another review?

dakshdeepHERE commented:
@SSwiniarski not yet give me this 3 days time. Im just a bit busy from last 2 months with jobs applications and tests. I am really sorry. Thanks for the reminder.

dakshdeepHERE commented:
Hey @HishamT, checkout the changes now. I have also removed the part where we print out the subarray to in the first example and shifted it to the codebytes part to keep things interesting. I have also changed the array as per your request and did the explanation too. I'm really sorry @SSwiniarski  for the delayed response and edits, won't happen again.

HishamT commented on content/general/concepts/algorithm/terms/euclidean-algorithm/kadane-algorithm.md:
```diff
@@ -0,0 +1,107 @@
+  - 'paths/computer-science'
+---
+
+**`Kadane's algorithm`** is a dynamic programming approach utilized to identify the maximum sum of a subarray within a given array of numbers. The algorithm solves this problem efficiently in a single pass through the array by maintaining two variables: `maxEndingHere` and `maxSoFar`. The algorithm iterates through the array, updating these variables at each step to track the maximum subarray sum ...
```
```suggestion
**`Kadane's algorithm`** is a dynamic programming approach to efficiently finding the maximum sum of a subarray in a given array of numbers. The algorithm works as follows:
1. Initialize a current sum (variable called `maxEndingHere`) equal to the value of the element at the first position in the array (`arr[0]`).
2. Iterate through the array. At every position, set `maxEndingHere` to the maximum of the following two values: `maxEndingHere + arr[i]` or `arr[i]`, where `i` is the current position in the array.
3. Keep track of the greatest `maxEndingHere` encountered (ex: in a variable called `maxSoFar`).
4. Return `maxSoFar`.
This algorithm guarantees a time complexity of `O(n)`, making it an optimal solution for this problem.
```

HishamT commented on content/general/concepts/algorithm/terms/euclidean-algorithm/kadane-algorithm.md:
```diff
@@ -0,0 +1,107 @@
+
+**Here's how Kadane's algorithm can be applied to a problem:**
+
+- Start with the first number and add it to a running total. Then, move to the next number and determine whether to continue with the current sum or start a new sum from this number. Always aim to maintain the largest sum seen so far.
```
If you accept my suggestion from above, you can delete this line.

HishamT commented on content/general/concepts/algorithm/terms/euclidean-algorithm/kadane-algorithm.md:
```diff
@@ -0,0 +1,107 @@
+**Here's how Kadane's algorithm can be applied to a problem:**
+
+- Start with the first number and add it to a running total. Then, move to the next number and determine whether to continue with the current sum or start a new sum from this number. Always aim to maintain the largest sum seen so far.
+- For example, consider the array {-2, -1, -3, 4, -1, 2, 1, -5, 4}. Initialize two integer variables: one named `maxEndingHere` and the other called `maxSoFar`. These variables serve to track the maximum subarray sum ending at the current position and the overall maximum subarray sum.
```
Put this in a section called "example". Use the ## Markdown header for this.

HishamT commented on content/general/concepts/algorithm/terms/euclidean-algorithm/kadane-algorithm.md:
```diff
@@ -0,0 +1,107 @@
+
+- Start with the first number and add it to a running total. Then, move to the next number and determine whether to continue with the current sum or start a new sum from this number. Always aim to maintain the largest sum seen so far.
+- For example, consider the array {-2, -1, -3, 4, -1, 2, 1, -5, 4}. Initialize two integer variables: one named `maxEndingHere` and the other called `maxSoFar`. These variables serve to track the maximum subarray sum ending at the current position and the overall maximum subarray sum.
+- The loop commence with the first element, which is -2. So, both `maxEndingHere` and `maxSoFar` start at -2.
```
```suggestion
The loop commences with the first element, which is -2. Therefore, both `maxEndingHere` and `maxSoFar` are initially -2.
```

HishamT commented on content/general/concepts/algorithm/terms/euclidean-algorithm/kadane-algorithm.md:
```diff
@@ -0,0 +1,107 @@
+- Start with the first number and add it to a running total. Then, move to the next number and determine whether to continue with the current sum or start a new sum from this number. Always aim to maintain the largest sum seen so far.
+- For example, consider the array {-2, -1, -3, 4, -1, 2, 1, -5, 4}. Initialize two integer variables: one named `maxEndingHere` and the other called `maxSoFar`. These variables serve to track the maximum subarray sum ending at the current position and the overall maximum subarray sum.
+- The loop commence with the first element, which is -2. So, both `maxEndingHere` and `maxSoFar` start at -2.
+- As the loop proceed to the next element, which is -1, compare it to the sum of the current element and the previous value of `maxEndingHere`. Update `maxEndingHere` to **-1** because it's greater than the sum of **-2** and **-1**. Also, compare `maxSoFar` with the updated value of `maxEndingHere`. Since **-2** is less than **-1**, update `maxSoFar` to **-1**.
```
```suggestion
Proceed to the next element, -1, and compare it to the sum of itself and the previous value of `maxEndingHere`.  Since -1 is greater than the sum of -2 and -1, set`maxEndingHere` to  -1. Also, compare `maxSoFar` with the updated value of `maxEndingHere`. Since -2 is less than -1 set `maxSoFar` to  -1.
```

HishamT reviewed (commented):
Reviewed

SSwiniarski commented:
@dakshdeepHERE you should address @HishamT's comments so we can get this PR moving.

dakshdeepHERE commented:
@SSwiniarski Yep I forgot about this one. Will complete it right away.

dakshdeepHERE commented:
Thank you @HishamT for the review. I guess we can merge it in coming days. ❤️🚀

SSwiniarski reviewed (changes requested):
@dakshdeepHERE, what you have is good. There's one problem though. Codebytes haven't been implemented for Java. So you need to change the codebyte to be a regular code block, or revise it into a programming language that is supported, like Python. (You can see the supported languages [here](https://www.codecademy.com/codebyte-editor) in the dropdown.)

dakshdeepHERE commented:
@SSwiniarski I can do it python and cpp. Let's say if I do that in python so do I have to add a note why I used Python? I guess then I will have to add learn-python in CatalogContent?

SSwiniarski commented:
> @SSwiniarski I can do it python and cpp. Let's say if I do that in python so do I have to add a note why I used Python? I guess then I will have to add learn-python in CatalogContent?

If you did both code examples in Python I don't think you'd need to add an explanation.

dakshdeepHERE commented:
@SSwiniarski I have changed the code snippets to cpp will it work now? The codebyte work for cpp too.

SSwiniarski reviewed (approved):
@dakshdeepHERE This looks good now.
